src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Infer.java

```diff
@@ -651,6 +654,91 @@ public Type instantiateFunctionalInterface(DiagnosticPosition pos, Type funcInte
             return owntype;
         }
     }
+
+    public Type instantiatePatternType(Type expressionType, TypeSymbol patternTypeSymbol) {
+        if (expressionType.tsym == patternTypeSymbol)
+            return expressionType;
+
+        //step 1:
+        Type expressionTypeCaptured = types.capture(expressionType);
+        List<Type> params = patternTypeSymbol.type.allparams();
+        List<Type> capturedWildcards = List.nil();
+        //add synthetic captured ivars
+        for (Type ta : expressionTypeCaptured.getTypeArguments()) {
+            if (ta.hasTag(TYPEVAR) && ((TypeVar)ta).isCaptured()) {
+                params = params.prepend((TypeVar)ta);
+                capturedWildcards = capturedWildcards.prepend(ta);
+            }
+        }
+        InferenceContext c = new InferenceContext(this, params);
+        Type patternType = c.asUndetVar(patternTypeSymbol.type);
+        Type exprType = c.asUndetVar(expressionTypeCaptured);
+
+        capturedWildcards.forEach(s -> ((UndetVar) c.asUndetVar(s)).setNormal());
+
+        try {
+            //step 2:
+            if (commonSuperWithDiffParameterization(patternType, exprType)) {
+                return null;
+            }
+
+            List<Type> varsToSolve = params.map(s -> c.asUndetVar(s));
+
+            doIncorporation(c, types.noWarnings);
+
+            while (c.solveBasic(varsToSolve, EnumSet.of(InferenceStep.EQ)).nonEmpty()) {
```

Comment:
if `varsToSolve` is equal to the number of variables in the inference context this step is a no-op and I think this should be always the case, so I think that the real resolution is happening below when `instantiatePatternVars` is invoked, unless I'm missing something this while loop could be safely removed